clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -18319,6 +18320,26 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
   return nullptr;
 }
 
+void CodeGenFunction::AddAMDGCNAddressSpaceMMRA(llvm::Instruction *Inst,
+                                                llvm::Value *ASMask) {
+  constexpr const char *Tag = "opencl-fence-mem";
+
+  uint64_t Mask = cast<llvm::ConstantInt>(ASMask)->getZExtValue();
+  if (Mask == 0)
+    return;
+
+  // 3 bits can be set: local, global, image in that order.
+  LLVMContext &Ctx = Inst->getContext();
+  SmallVector<MMRAMetadata::TagT, 3> MMRAs;
+  if (Mask & (1 << 0))
```

Comment:
These magic numbers for the shift amount should be declared as an Enum somewhere.

Reply:
Agreed, but not sure where they should go to be meaningful. Any ideas?
Maybe `TargetParser` could have a list of MMRA-related values (tag prefix/suffixes, mask values for builtins, etc.) ?

Reply:
Alternatively could behave more like system scopes where the builtin takes a variadic list of string literal arguments passed directly through

Reply:
@arsenm Do you mean just having a space-separate list of address spaces such as `"image local"`?
If that's acceptable then it definitely works for me too